codex/slots/proofs/backends/asynccircoms.nim

```diff
@@ -0,0 +1,82 @@
+import std/options
+
+import pkg/taskpools
+import pkg/taskpools/flowvars
+import pkg/chronos
+import pkg/chronos/threadsync
+import pkg/questionable/results
+
+import ../../types
+import ../../../utils/asyncthreads
+
+import ./circomcompat
+
+type
+  AsyncCircomCompat* = object
+    params*: CircomCompatParams
+    tp*: Taskpool
+
+  # Args objects are missing seq[seq[byte]] field, to avoid unnecessary data copy
+  ProveTaskArgs* = object
+    signal: ThreadSignalPtr
+    params: CircomCompatParams
+
+var circomBackend {.threadvar.}: Option[CircomCompat]
+
+proc proveTask[H](
+    args: ProveTaskArgs, data: ProofInputs[H]
+): Result[CircomProof, string] =
+
+  try:
+    if circomBackend.isNone:
+      circomBackend = some CircomCompat.init(args.params)
```

Comment:
Not a big fan of this pattern, I would leave initialization to the caller.

Reply:
It is pretty awkward. I was trying to avoid having to make a new instance on each call as that seems overkill. 

Though, the backend *should* be threadsafe so I could switch it to use one global config since the Rust code clones it anyways.

I'll switch it out with one global config.

Reply:
Though one reason to do this on each worker thread would be to avoid blocking the Chronos thread while CircomCompat loads the zkey. Though that currently runs pretty quickly right?

Reply:
We don't need an instance per thread, internally circom compat should be thread safe. So just create one instance and pass (a pointer to) it to the task on invocation.

> Though one reason to do this on each worker thread would be to avoid blocking the Chronos thread while CircomCompat loads the zkey. Though that currently runs pretty quickly right?

This only happens once on startup and it shouldn't be a big deal. If it becomes an issue, we can move initialization to is't own thread as well.

Reply:
> This only happens once on startup and it shouldn't be a big deal. If it becomes an issue, we can move initialization to is't own thread as well.

Okay, yah that avoids a bit of complication.